### ISSUE TITLE
WhatsApp messages sending and reading

### ISSUE THREAD

**Issue by lc-at** [p4kl0nc4t/kyros#2]

**Comment by irvanrahadhian** [comment]
Hello, thank for this amazing lib, but i still cannot sending message, i tried to follow the simple example. already logged in, but still cant sending message.

**Comment by lc-at** [comment]
Sending messages is currently still a pending task. In order to send a message, the binary websocket message decoding process still has to be done in the first place (ongoing).

**Comment by RynKings** [comment]
This lib isn't continued yet ?

**Comment by lc-at** [comment]
> This lib isn't continued yet ?

Yes, it is still pending. If you want, you can contribute and help develop the feature. I will be very welcome to see your pull request.

**Comment by MaverickAlonso** [comment]
Hi guys, I saw that this feature is still pending. Unfortunately, I do know what should be the next steps after the first part. Do we know which steps need to be developed for sending? Do we have a plan or details? maybe we can help in advance with this future.  Regards,

**Comment by lc-at** [comment]
Sorry for the lagging development. I really wish I could work on it sooner. If you want to work on the feature, you can follow the reference by looking at [whatsapp-web-reveng](https://github.com/sigalor/whatsapp-web-reveng) or [go-whatsapp](https://github.com/Rhymen/go-whatsapp) (especially the part of code where they do binary encoding and decoding). So far, we have been able to decrypt the websocket messages. From all the websocket messages, some are binary messages and some are just regular string. Before being able to send and read WhatsApp message, I believe that the first thing we have to work out is the binary decoder/encoder (#3).

Regards,

**Comment by p4kl0nc4t** [comment]
#3

**Comment by RezSat** [comment]
see https://github.com/ravindu01manoj/Sew-Queen they have built a complete WhatsApp bot, and every easy deploy with the direct link to Heroku. But it's done with Javascript or rather Node-JS, we might able to reimplement it in python as well.

**Comment by lc-at** [comment]
@RezSat I will take a look on that.

**Comment by RezSat** [comment]
> @RezSat I will take a look on that.

I think it will be harder because they haven't completely open sourced it, some of the features aren't readable on normal manner.

**Comment by reinier-millo** [comment]
I have added to develop branch the binary reader for WhatsApp messages. Currently the implementation is receiving text plain messages. To integrate with the message handler i have added a message callback. Changes are uploaded with #21.

There is some work to do with the received message to identify different type of messages. Currently media messages are not decoding successfully. Decoding is failing with the protobuf decode. I'm looking some references to identify how to solve it and allow to receive media messages. Currently media messages are received but don't decoded and identified successfully.
